README.md
## Неофициальный ЕЭТК телеграм бот 

Телеграм бот, который уведомляет об изменениях на сайте ЕЭТК

Список основных файлов:
- collect_data_daily.py — Отвечает за парсинг и скачивание ежедневных изменений
- collect_data_weekly.py — Отвечает за парсинг и скачивание еженедельных расписаний для всех курсов
- collect_kitties.py — Отвечает за милых котиков :)
- eetkbot.py — **Исполняемый файл** телеграм бота
- main.py — **Исполняемый файл** отлова любых изменений на сайте

Как уже можно было понять, для корректной работы нужно запускать 2 файла - eetkbot.py и main.py

Однако по одиночке они тоже работают

Внутри каждого файла присутствуют комментарии о том как все работает и для чего это нужно

Поднимать бота можно как на windows, так и на linux. Лично я использовал VDS на ОС ubuntu.

### Инструкция для Linux:
После подключения к VDS вводим в терминал следующее:
```
Установим Python версии не ниже 3.7 (самостоятельно)
sudo apt install python3-venv
```
Загрузим на наш выделенный сервер файлы проекта. Это можно сделать как с помощью различных файловых менеджеров, с помощью которых вы подключились, так и просто клонируя этот репозиторий. Ничего сложного :)

Перейдем в папку с проектом на выделенном сервере и создадим виртуальное окружение
```
cd "имя папки"
python3 -m venv venv
source venv/bin/activate
```
Обновим пакетный менеджер и установим необходимые для работы библиотеки
```
pip install -U pip
pip install lxml requests fake-useragent beautifulsoup4 aiogram google-search-results pydantic
```
Ну вот и всё. Нужные библиотеки установлены. Теперь мы можем запускать бота

Однако чтобы он работал даже без вашего присутствия необходимо запустить его в отдельном экране с помощью утилиты screen. Если она не установлена по умолчанию, то скачать ее можно командой ***apt install screen***

Про использование screen хорошо расписано [здесь](https://wiki.merionet.ru/servernye-resheniya/40/kak-polzovatsya-utilitoj-screen-v-linux/)

Создадим два экрана для двух исполняемых файлов
```
screen -S main
screen -S eetkbot
```
С помощью команды ***screen -r main*** зайдем в экран и запустим в нём исполняемый файл
```
python3.7 main.py
```
Чтобы остановить работу скрипта нажимаем Ctrl+C

Чтобы свернуть экран нажимаем Ctrl+A и D

Аналогичным образом во втором экране запускаем второй исполняемый файл

### Инструкция для Windows:
Установим Python с [официального сайта](https://www.python.org/downloads/) версии не ниже 3.7

При установке обязательно ставим галочку на **Add Python to PATH**

Загрузим файлы проекта и поместим их в одну папку

Открываем cmd или PowerShell Terminal в этой папке и создадим виртуальное окружение
```
python -m venv venv
.\.venv\Scripts\activate
```
Теперь загрузим необходимые библиотеки
```
pip install lxml requests fake-useragent beautifulsoup4 aiogram google-search-results pydantic
```
Для каждого из исполняемых файлов нужно своё окно командной строки. Это значит, что нужно открыть еще одно окно cmd или PowerShell Terminal и активировать в нём виртуальное окружение
```
.\.venv\Scripts\activate
```
Запустим исполняемый файл командой ***python main.py***

Аналогичным образом во втором окне запускаем второй исполняемый файл

## Возможные проблемы при запуске
```
UnicodeEncodeError: 'latin-1' codec can't encode characters in position 0-7: ordinal not in range(256)
```
Такая ошибка означает, что в вашем VDS не установлена RU локализация, которая необходима для работы.

Решение здесь: https://qna.habr.com/q/684410 (комментарий @CopsisOnFire), плюсом к этому нужно прописать в рабочем экране ***LANG=ru_RU.UTF***




texts/changes.txt
https://cloud.mail.ru/public/sCbB/4Anem4a7d
http://eetk.ru/wp-content/uploads/2022/03/izmeneniya-studentov-na-25-marta-pyatnicza-1.pdf
http://eetk.ru/wp-content/uploads/2022/03/izmeneniya-studentov-na-26-marta-subbota.pdf


main.py
from collect_data_weekly import collect_data_weekly
from collect_data_daily import collect_data_daily
from fake_useragent import UserAgent
from bs4 import BeautifulSoup
import requests
import time
import glob
import os

# Ввод глобальных переменных
data_weekly = []
data_daily = []
result_weekly = []
result_daily = []
user_id = []
# Формирование фейкового юзер-агента, чтобы сайт не блочил доступ при частых запросах
ua = UserAgent()

# Вводим токен нашего бота с припиской bot
token = 'bot5085326595:AAGVsDbtRcsj4at6haoV10d5_vSnBoaeNqg'

# 5068878742:AAHeLzjo7WNXb_siW9YVQTKpfae3R6wKHIg Токен главного бота
# 5085326595:AAGVsDbtRcsj4at6haoV10d5_vSnBoaeNqg Токен тестового бота

# 760196701 Это я)) @lxvko


# Собираем актуальные данные для использования в качестве оригинала
def collect_data():
    global data_weekly
    data_weekly.clear()
    r = requests.get(
        url='http://eetk.ru/78-2/82-2/88-2/',
        headers={'user-agent': f'{ua.random}'})

    soup = BeautifulSoup(r.text, "lxml")
    schedules = soup.find_all("tr", class_="row-4 even")
    for schedule in schedules:
        data_weekly += [link.get("href") for link in schedule.find_all("a")]

    global data_daily
    data_daily.clear()
    r = requests.get(
        url='http://eetk.ru/78-2/2953-2/3115-2/',
        headers={'user-agent': f'{ua.random}'})

    soup = BeautifulSoup(r.text, "lxml")
    changes = soup.find_all("article", class_="post-3115 page type-page status-publish hentry")
    for change in changes:
        data_daily += [link.get("href") for link in change.find_all("a")]

    collect_result()


# Проверяем на наличие отличий в вечном цикле
def collect_result():
    while True:
        # Такие большие задержки нужны чтобы парсинг был 24/7 и чтобы сайт не запрещал доступ

        time.sleep(40)

        if collect_result_weekly() is False:
            continue

        time.sleep(40)

        if collect_result_daily() is False:
            continue

        # Сверяем оригинал с копией (первая проверка)
        if data_weekly != result_weekly:
            check(result_weekly)
        elif data_daily != result_daily:
            check(result_daily)
        # Если все совпало, то очищаем копии
        else:
            result_weekly.clear()
            result_daily.clear()


# Проверка расписания
def collect_result_weekly():
    # Собираем актуальные данные для использования в качестве копии
    global result_weekly
    result_weekly.clear()

    # Обход различных ошибок со стороны сайта
    try:
        r = requests.get(
            url='http://eetk.ru/78-2/82-2/88-2/',
            headers={'user-agent': f'{ua.random}'})
    except requests.ConnectionError as e:
        print(str(e))
        time.sleep(60)
        return False
    except requests.Timeout as e:
        print(str(e))
        time.sleep(60)
        return False
    except requests.RequestException as e:
        print(str(e))
        time.sleep(60)
        return False
    except KeyboardInterrupt:
        print("Something closed the program")

    soup = BeautifulSoup(r.text, "lxml")
    schedules = soup.find_all("tr", class_="row-4 even")
    for schedule in schedules:
        result_weekly += [link.get("href") for link in schedule.find_all("a")]
    return True


# Проверка измерений
def collect_result_daily():
    # Собираем актуальные данные для использования в качестве копии
    global result_daily
    result_daily.clear()

    # Обход различных ошибок со стороны сайта
    try:
        r = requests.get(
            url='http://eetk.ru/78-2/2953-2/3115-2/',
            headers={'user-agent': f'{ua.random}'})
    except requests.ConnectionError as e:
        print(str(e))
        time.sleep(60)
        return False
    except requests.Timeout as e:
        print(str(e))
        time.sleep(60)
        return False
    except requests.RequestException as e:
        print(str(e))
        time.sleep(60)
        return False
    except KeyboardInterrupt:
        print("Something closed the program")
        return False

    soup = BeautifulSoup(r.text, "lxml")
    changes = soup.find_all("article", class_="post-3115 page type-page status-publish hentry")
    for change in changes:
        result_daily += [link.get("href") for link in change.find_all("a")]
    return True


# Вторая сверка оригинала с копией (на вход подается копия)
def check(check_data):
    if check_data == result_weekly:
        time.sleep(30)
        # Собираем актуальные данные для использования в качестве копии
        collect_result_weekly()
        # Ожидаем время и еще раз проводим сверку оригинала с копией
        if data_weekly != result_weekly:
            # Если на сайте точно есть новое расписание - запускаем рассылку
            send_message('На сайте выложили новое расписание!')
            # Производим обновление папки pdfs до актуальной версии
            reload_data()
            # Получаем новый оригинал и отправляемся крутить цикл дальше
            collect_data()
    elif check_data == result_daily:
        time.sleep(30)
        # Собираем актуальные данные для использования в качестве копии
        collect_result_daily()
        # Ожидаем время и еще раз проводим сверку оригинала с копией
        if data_daily != result_daily:
            # Если на сайте точно есть изменения - запускаем рассылку
            send_message('На сайте появились изменения!')
            # Производим обновление папки pdfs до актуальной версии
            reload_data()
            # Получаем новый оригинал и отправляемся крутить цикл дальше
            collect_data()


# Функция отправки сообщений участникам рассылки
def send_message(message):
    global user_id
    user_id.clear()

    # Подгружаем базу данных пользователей
    with open('user_id.txt') as file:
        user_id = [line.strip() for line in file]

    # Отправляем каждому пользователю по очереди сообщение с нужным текстом
    for req in range(len(user_id)):
        time.sleep(1)
        r = requests.get(f'https://api.telegram.org/{token}/sendMessage'
                         f'?chat_id={user_id[req]}&text={message}')


# Обновление папки pdfs до актуальной версии
def reload_data():
    # Удаляем все файлы из папки pdfs
    for file in glob.glob('pdfs/*'):
        os.remove(file)

    # Скачиваем актуальные файлы изменений
    collect_data_daily()

    course = 1
    # Скачиваем актуальные файлы расписания для всех курсов
    for course in range(1, 5):
        time.sleep(2)
        try:
            collect_data_weekly(course)
        except:
            pass


if __name__ == '__main__':
    collect_data()


eetkbot.py
from aiogram import Bot, Dispatcher, executor, types
from aiogram.dispatcher.filters import Text
from collect_data_weekly import collect_data_weekly
from collect_data_daily import collect_data_daily
from collect_kitties import collect_kitty
import asyncio
import glob
import os


# Вводим токен нашего бота
bot = Bot(token='5085326595:AAGVsDbtRcsj4at6haoV10d5_vSnBoaeNqg', parse_mode=types.ParseMode.HTML)
dp = Dispatcher(bot)


# 5068878742:AAHeLzjo7WNXb_siW9YVQTKpfae3R6wKHIg Токен главного бота
# 5085326595:AAGVsDbtRcsj4at6haoV10d5_vSnBoaeNqg Токен тестового бота

# 760196701 Это я)) @lxvko

# Обработчик команды /start
@dp.message_handler(commands='start')
async def start(message: types.Message):
    # Создание кнопок для выбора курсов
    start_buttons = ['Первый курс', 'Второй курс', 'Третий курс', 'Четвертый курс']
    keyboard = types.ReplyKeyboardMarkup(resize_keyboard=True)
    for start_button in start_buttons:
        keyboard.add(start_button)

    # Отправка сообщения пользователю
    await message.answer('Выбери нужный курс', reply_markup=keyboard)


# Обработчик команды "Вернуться к выбору курсов"
@dp.message_handler(Text(equals='Вернуться к выбору курсов'))
async def comeback(message: types.Message):
    # Создание кнопок для выбора курсов
    start_buttons = ['Первый курс', 'Второй курс', 'Третий курс', 'Четвертый курс']
    keyboard = types.ReplyKeyboardMarkup(resize_keyboard=True)
    for start_button in start_buttons:
        keyboard.add(start_button)

    # Отправка сообщения пользователю
    await message.answer('Выбери нужный курс', reply_markup=keyboard)


# Обработчик команды "/join"
@dp.message_handler(commands=['join'])
async def join(message: types.Message):
    # Бот берет id пользователя, написавшего команду
    current_user_id = message.from_user.id
    # Подгружает базу данных пользователей рассылки из файла
    with open('user_id.txt') as file:
        user_id = [line.strip() for line in file]

    # Если такой пользователь уже есть в базе данных
    if str(current_user_id) in user_id:
        await message.answer('Ты уже подписан на рассылку!')
    # Если нет в базе данных
    else:
        await message.answer('Теперь ты подписан на рассылку 😊\nЕсли захочешь отписаться, то напиши /leave')
        user_id.append(current_user_id)

    # Сохраняет базу данных пользователей, подписавшихся на рассылку
    with open("user_id.txt", "w") as file:
        for user in user_id:
            print(user, file=file)


# Обработчик команды "/leave"
@dp.message_handler(commands=['leave'])
async def leave(message: types.Message):
    # Бот берет id пользователя, написавшего команду
    current_user_id = message.from_user.id
    # Подгружает базу данных пользователей рассылки из файла
    with open('user_id.txt') as file:
        user_id = [line.strip() for line in file]

    # Пробует удалить пользователя из базы данных
    try:
        user_id.remove(str(current_user_id))
        await message.answer('Ты отписался от рассылки 🥺')
    except ValueError:
        await message.answer('Но ведь тебя и так нет в рассылке!')

    # Сохраняет базу данных
    with open("user_id.txt", "w") as file:
        for user in user_id:
            print(user, file=file)


# Обработчик секретной команды "/spisok"
@dp.message_handler(commands=['spisok'])
async def spisok(message: types.Message):
    # Открываем базу данных пользователей и выводим их в чат с пользователем
    with open('user_id.txt') as file:
        user_id = [line.strip() for line in file]

    for i in range(len(user_id)):
        await message.answer(f'{user_id[i]}')


# Обработчик секретной команды "/preload"
@dp.message_handler(commands=['preload'])
async def spisok(message: types.Message):
    await message.answer('Загрузка. Подожди, пожалуйста 🙃')

    # Удаляем все файлы из папки pdfs
    for file in glob.glob('pdfs/*'):
        os.remove(file)

    # Скачиваем актуальные файлы изменений
    collect_data_daily()
    course = 1
    # Скачиваем актуальные файлы расписания для всех курсов
    for course in range(1, 5):
        await asyncio.sleep(2)
        try:
            collect_data_weekly(course)
            await message.answer(f'{course} курс успешно загружен 🙃')
        except:
            await message.answer(f'Ой, что-то сломалось с {course} курсом :(')


# Тут просто копипаст с минимальными изменениями для каждого курса, 
@dp.message_handler(Text(equals='Первый курс'))
async def first_course_is_selected(message: types.Message):
    first_course = ['Расписание первого курса', 'Изменения']
    keyboard = types.ReplyKeyboardMarkup(resize_keyboard=True)
    keyboard.add(*first_course)
    keyboard.add('Вернуться к выбору курсов')
    await message.answer('Теперь сделай запрос', reply_markup=keyboard)


@dp.message_handler(Text(equals='Второй курс'))
async def second_course_is_selected(message: types.Message):
    second_course = ['Расписание второго курса', 'Изменения']
    keyboard = types.ReplyKeyboardMarkup(resize_keyboard=True)
    keyboard.add(*second_course)
    keyboard.add('Вернуться к выбору курсов')
    await message.answer('Теперь сделай запрос', reply_markup=keyboard)


@dp.message_handler(Text(equals='Третий курс'))
async def third_course_is_selected(message: types.Message):
    third_course = ['Расписание третьего курса', 'Изменения']
    keyboard = types.ReplyKeyboardMarkup(resize_keyboard=True)
    keyboard.add(*third_course)
    keyboard.add('Вернуться к выбору курсов')
    await message.answer('Теперь сделай запрос', reply_markup=keyboard)


@dp.message_handler(Text(equals='Четвертый курс'))
async def fourth_course_is_selected(message: types.Message):
    fourth_course = ['Расписание четвертого курса', 'Изменения']
    keyboard = types.ReplyKeyboardMarkup(resize_keyboard=True)
    keyboard.add(*fourth_course)
    keyboard.add('Вернуться к выбору курсов')
    await message.answer('Теперь сделай запрос', reply_markup=keyboard)


# Копипаст с минимальными изменениями для каждого курса
@dp.message_handler(Text(equals='Расписание первого курса'))
async def get_data_weekly_first(message: types.Message):
    await message.answer('Загрузка. Подожди, пожалуйста 🙃')

    # Сохраняем pdf-файлы для первого курса
    collect_data_weekly(1)
    # Открываем названия файлов и ссылки
    with open('texts/schedules.txt') as file:
        schedule = [line.strip() for line in file]
    with open('texts/schedules_names.txt') as file:
        schedule_name = [line.strip() for line in file]

    # Отправляем файлы пользователю
    for i in range(len(schedule)):
        part = schedule[i].rpartition('/')[-1]
        await message.answer_document(
            open(f'pdfs/{part}.pdf', 'rb'),
            caption=f'{schedule_name[i]}')
        await asyncio.sleep(0.5)


@dp.message_handler(Text(equals='Расписание второго курса'))
async def get_data_weekly_second(message: types.Message):
    await message.answer('Загрузка. Подожди, пожалуйста 🙃')

    collect_data_weekly(2)
    with open('texts/schedules.txt') as file:
        schedule = [line.strip() for line in file]
    with open('texts/schedules_names.txt') as file:
        schedule_name = [line.strip() for line in file]

    for i in range(len(schedule)):
        part = schedule[i].rpartition('/')[-1]
        await message.answer_document(
            open(f'pdfs/{part}.pdf', 'rb'),
            caption=f'{schedule_name[i]}')
        await asyncio.sleep(0.5)


@dp.message_handler(Text(equals='Расписание третьего курса'))
async def get_data_weekly_third(message: types.Message):
    await message.answer('Загрузка. Подожди, пожалуйста 🙃')

    collect_data_weekly(3)
    with open('texts/schedules.txt') as file:
        schedule = [line.strip() for line in file]
    with open('texts/schedules_names.txt') as file:
        schedule_name = [line.strip() for line in file]

    for i in range(len(schedule)):
        part = schedule[i].rpartition('/')[-1]
        await message.answer_document(
            open(f'pdfs/{part}.pdf', 'rb'),
            caption=f'{schedule_name[i]}')
        await asyncio.sleep(0.5)


@dp.message_handler(Text(equals='Расписание четвертого курса'))
async def get_data_weekly_fourth(message: types.Message):
    await message.answer('Загрузка. Подожди, пожалуйста 🙃')

    collect_data_weekly(4)
    with open('texts/schedules.txt') as file:
        schedule = [line.strip() for line in file]
    with open('texts/schedules_names.txt') as file:
        schedule_name = [line.strip() for line in file]

    for i in range(len(schedule)):
        part = schedule[i].rpartition('/')[-1]
        await message.answer_document(
            open(f'pdfs/{part}.pdf', 'rb'),
            caption=f'{schedule_name[i]}')
        await asyncio.sleep(0.5)


# Обработчик команды "Изменения"
@dp.message_handler(Text(equals='Изменения'))
async def get_data_daily(message: types.Message):
    await message.answer('Загрузка. Подожди, пожалуйста 🙃')

    # Сохраняем pdf-файлы изменений
    collect_data_daily()
    # Подгружаем имена файлов, ссылки и форматы файлов
    with open('texts/changes.txt') as file:
        changes = [line.strip() for line in file]
    with open('texts/changes_names.txt') as file:
        changes_names = [line.strip() for line in file]
    with open('texts/temp_formats.txt') as file:
        temp_formats = [line.strip() for line in file]

    counter = -1
    for change in changes:
        counter += 1
        await asyncio.sleep(0.5)
        # Отправляем файлы пользователю с интервалом пол секунды
        if 'cloud.mail.ru' in change:
            part = changes[counter].rpartition('/')[-1]
            if temp_formats[counter] == 'pdf':
                await message.answer_document(
                    open(f'pdfs/{part}.pdf', 'rb'),
                    caption=f'{changes_names[counter]}')
            elif temp_formats[counter] == 'jpg':
                await message.answer_document(
                    open(f'pdfs/{part}.jpg', 'rb'),
                    caption=f'{changes_names[counter]}')
            elif temp_formats[counter] == 'png':
                await message.answer_document(
                    open(f'pdfs/{part}.png', 'rb'),
                    caption=f'{changes_names[counter]}')
        else:
            await message.answer_document(
                open(f'pdfs/{changes_names[counter]}.pdf', 'rb'),
                caption=f'{changes_names[counter]}')


# Обработчик секретной команды "Милый котик"
@dp.message_handler(Text(equals='Милый котик'))
async def get_kitty(message: types.Message):
    # Сохраняем котика
    collect_kitty()
    # Отправляем котика
    await message.answer_photo(open('cats/kitty.jpg', 'rb'))


# Запускаем бота
def main():
    executor.start_polling(dp)


if __name__ == '__main__':
    main()


collect_kitties.py
# Эта фича была придумана и реализована будучи в нетрезвом состоянии ума и тела
# О ней почти никто не знает и на том спасибо :)
# Кодовое слово в ТГ-боте "Милый котик"

# По-хорошему, реализацию фичи нужно переделать под сайт 'https://random.cat/'
# Но мне лень

from serpapi import GoogleSearch
import requests
import random
import json


# Функция обращается к сервису, который собирает картинки по параметру "q":
def get_google_images():
    params = {
        "api_key": "d094df85fa3f2b1047d7f464e3989d9c48ccbba4bd4359b982e10333ddf1a89a",
        "engine": "google",
        "ijn": "0",
        "q": "милые котики",
        "google_domain": "google.com",
        "tbm": "isch"
    }

    search = GoogleSearch(params)
    results = search.get_dict()
    # Записываем 100 первых ссылок на картинки с котиками из гугла
    with open('cats/data.json', 'w') as file:
        file.write(json.dumps(results['images_results'], indent=2))


# Получаем котиков в картинках (На вход подается json котиков)
def get_kitty(kitties):
    counter = []
    # Копируем оригинальный json для внесения правок
    kitties_work = kitties.copy()
    for i in reversed(range(len(kitties))):
        # Перечень сайтов, которые так или иначе плохо возвращают котиков (список можно пополнять)
        if \
                kitties_work[i]['source'] == 'm.facebook.com' \
                or kitties_work[i]['source'] == 'facebook.com' \
                or kitties_work[i]['source'] == 'ont.by' \
                or kitties_work[i]['source'] == 'novosti-n.org' \
                or kitties_work[i]['source'] == 'ysia.ru' \
                or kitties_work[i]['source'] == 'krasivosti.pro' \
                or kitties_work[i]['source'] == 'mover.uz' \
                or kitties_work[i]['source'] == 'cultmall.com.ua · In stock' \
                or kitties_work[i]['source'] == 'mariart.kiev.ua · In stock' \
                or kitties_work[i]['source'] == 'pets.24tv.ua' \
                or kitties_work[i]['source'] == 'stickers.wiki':
            # Эти сайты удаляются из json с котиками, чтобы в будущем не мешать работе
            del kitties_work[i]
    # Реализуем номера для котиков
    for i in range(len(kitties_work)):
        counter.append(kitties_work[i]['position'])

    # Берем рандомный номер из списка
    count = random.choice(counter)
    for kitty in kitties_work:
        if kitty.get('position') == count:
            # Сохраняем котика :)
            image = requests.get(kitty['original']).content
            # Удаляем использованного котика :(
            del kitties_work[counter.index(count)]
            counter.remove(count)

    # Записываем котика в файл
    with open(f'cats/kitty.jpg', 'wb') as file:
        file.write(image)

    # Перезаписываем базу данных котиков
    with open('cats/data.json', 'w') as file:
        file.write(json.dumps(kitties_work, indent=2))


# Основная функция
def collect_kitty():
    # Открываем базу данных котиков из файла
    with open('cats/data.json', 'r') as file:
        kitties = json.load(file)
    # Если в базе нет ни одного котика, то парсим новых через сервис
    if len(kitties) == 0:
        get_google_images()
        main()
    # Иначе сохраняем котика на компьютер
    else:
        get_kitty(kitties)


def main():
    collect_kitty()


if __name__ == '__main__':
    main()


collect_data_weekly.py
from fake_useragent import UserAgent
from pydantic import BaseModel
from bs4 import BeautifulSoup
import requests
import os

# Формирование фейкового юзер-агента, чтобы сайт не блочил доступ при частых запросах
ua = UserAgent()


# Упрощенная система парсинга ссылок для скачивания pdf-файлов с cloud.mail.ru
class Format(BaseModel):
    name: str


class Link(BaseModel):
    url: str


# Сбор данных с сайта колледжа (Промежуток расписания и ссылка на него)
# На вход поступает номер курса из ТГ-бота
def collect_data_weekly(course):
    data_weekly = []
    date_weekly = []

    r = requests.get(
        url='http://eetk.ru/78-2/82-2/88-2/',
        headers={'user-agent': f'{ua.random}'})

    soup = BeautifulSoup(r.text, "lxml")

    if course == 1:
        schedules = soup.find_all("tr", class_="row-1 odd")
    elif course == 2:
        schedules = soup.find_all("tr", class_="row-2 even")
    elif course == 3:
        schedules = soup.find_all("tr", class_="row-3 odd")
    elif course == 4:
        schedules = soup.find_all("tr", class_="row-4 even")

    schedules_names = soup.find_all(
        "article", class_="post-88 page type-page status-publish hentry")

    for schedule in schedules:
        data_weekly += [link.get("href") for link in schedule.find_all("a")]
    for schedules_name in schedules_names:
        date_weekly += [link.text for link in schedules_name.find_all("h3")]

    # Записываем в файл ссылки на расписание
    with open("texts/schedules.txt", "w") as file:
        for item in data_weekly:
            print(item, file=file)
    # Записываем в файл промежуток расписания
    with open("texts/schedules_names.txt", "w") as file:
        for item in date_weekly:
            print(item, file=file)

    download_file(data_weekly)


# Скачивание файлов для отправки в ТГ-бота (на вход подаются ссылки)
def download_file(ids):
    for id in ids:

        part = id.rpartition('public/')[-1]  # Кусок будущей ссылки для скачивания
        name = id.rpartition('/')[-1]  # Будущее имя файла

        # Если файл уже скачан, то еще раз он качаться не будет
        if os.path.isfile(f'pdfs/{name}.pdf'):
            continue

        # Парсим сайт cloud.mail.ru
        temp = requests.get(id)
        temp = temp.text

        # Магическим образом достаем ссылку для скачивания
        magic_link = Link.parse_raw(temp.partition('"weblink_get": ')[2].partition(',\n\t\t"')[0])
        pdf = requests.get(magic_link.url + f'/{part}')

        # Скачиваем нужный нам pdf-файл
        with open(f'pdfs/{name}.pdf', 'wb') as f:
            f.write(pdf.content)


def main():
    collect_data_weekly()


if __name__ == '__main__':
    main()


collect_data_daily.py
from fake_useragent import UserAgent
from urllib.parse import urlencode
from pydantic import BaseModel
from bs4 import BeautifulSoup
import requests
import os

# Формирование фейкового юзер-агента, чтобы сайт не блочил доступ при частых запросах
ua = UserAgent()


# Упрощенная система парсинга ссылок для скачивания pdf-файлов с cloud.mail.ru
class Format(BaseModel):
    name: str


class Link(BaseModel):
    url: str


# Сбор данных с сайта колледжа (Имя изменений и ссылка на них)
def collect_data_daily():
    data_daily = []
    date_daily = []

    r = requests.get(
        url='http://eetk.ru/78-2/2953-2/3115-2/',
        headers={'user-agent': f'{ua.random}'})

    soup = BeautifulSoup(r.text, "lxml")
    changes = soup.find_all(
        "article", class_="post-3115 page type-page status-publish hentry")
    changes_names = changes

    for change in changes:
        data_daily += [link.get("href") for link in change.find_all("a")]
    for changes_name in changes_names:
        date_daily += [link.text for link in changes_name.find_all("a")]

    # Записываем в файл ссылки на изменения
    with open("texts/changes.txt", "w") as file:
        for item in data_daily:
            print(item, file=file)
    # Записываем в файл имена изменений
    with open("texts/changes_names.txt", "w") as file:
        for item in date_daily:
            print(item, file=file)

    download_file(data_daily, date_daily)


# Скачивание файлов для отправки в ТГ-бота (на вход подаются ссылки и имена)
def download_file(ids, names):
    counter = -1
    temp_formats = []
    for id in ids:
        counter += 1
        # Если pdf-файл залит в cloud.mail.ru
        if 'cloud.mail.ru' in id:
            try:
                part = id.rpartition('public/')[-1]  # Кусок будущей ссылки для скачивания
                name = id.rpartition('/')[-1]  # Будущее имя файла

                # Парсим сайт cloud.mail.ru
                temp = requests.get(id)
                temp = temp.text

                # Магическим образом достаем формат будущего файла
                magic_format = Format.parse_raw(temp.partition('"list": [\n')[2].partition(']')[0])
                temp_formats.append(magic_format.name.split('.')[-1])

                # Магическим образом достаем ссылку для скачивания
                magic_link = Link.parse_raw(temp.partition('"weblink_get": ')[2].partition(',\n\t\t"')[0])
                pdf = requests.get(magic_link.url + f'/{part}')

                # Если файл уже скачан, то еще раз он качаться не будет
                if os.path.isfile(f'pdfs/{name}.pdf'):
                    continue
                # Проверка на jpg формат.. :/ Были случаи..
                elif os.path.isfile(f'pdfs/{name}.jpg'):
                    continue

                # Скачиваем нужный нам pdf-файл
                with open(f'pdfs/{name}.{temp_formats[-1]}', 'wb') as f:
                    f.write(pdf.content)

            # Отлов ошибок при скачивании, которые будут выводиться в консоль
            except Exception as ex:
                print(f'Не удалось скачать файл {id}, из-за {ex}')

        # Если pdf-файл залит на Яндекс Диск (боже, упаси)
        # Возможно ТГ-бот их не отправит, потому что мне было лень реализовывать НАСТОЛЬКО редкий случай :)
        elif 'disk.yandex.ru' in id:
            try:
                # Это база (формируем запрос для получения ссылки скачивания с Яндекс Диска)
                base_url = 'https://cloud-api.yandex.net/v1/disk/public/resources/download?'
                final_url = base_url + urlencode(dict(public_key=id))

                # Обращаемся к API сайта
                response = requests.get(final_url)
                download_url = response.json()['href']  # Вытаскиваем ссылку на скачивание pdf-файла
                name = (download_url.split('&filename='))[1].split('&disposition')[0]  # Вытаскиваем имя будущего файла
                temp_formats.append(name.split('.')[-1])

                # Если файл уже скачан, то еще раз он качаться не будет
                if os.path.isfile(f'pdfs/{name}'):
                    continue

                # Скачиваем нужный нам pdf-файл
                pdf = requests.get(download_url)
                with open(f'pdfs/{name}', 'wb') as f:
                    f.write(pdf.content)

            # Отлов ошибок при скачивании, которые будут выводиться в консоль
            except Exception as ex:
                print(f'Не удалось скачать файл {id}, из-за {ex}')

        # Если pdf-файл залит на сайт колледжа
        else:
            # Если файл уже скачан, то еще раз он качаться не будет
            if os.path.isfile(f'pdfs/{names[counter]}.pdf'):
                continue

            # Скачиваем нужный нам pdf-файл
            response = requests.get(id)
            temp_formats.append('pdf')
            with open(f'pdfs/{names[counter]}.pdf', 'wb') as f:
                f.write(response.content)

    # Ведем запись форматов, которые сейчас находятся в папке, для корректной работы ТГ-бота
    if temp_formats:
        with open('texts/temp_formats.txt', 'w') as file:
            for item in temp_formats:
                print(item, file=file)


def main():
    collect_data_daily()


if __name__ == '__main__':
    main()
